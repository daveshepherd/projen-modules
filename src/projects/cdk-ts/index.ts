import {
  AwsCdkTypeScriptApp,
  AwsCdkTypeScriptAppOptions,
} from 'projen/lib/awscdk';
import { CdkTypeScriptAppOptions } from './cdk-typescript-app-options';
import { CodeOwners } from '../../components/github/codeowners';
import { DEFAULT_PULL_REQUEST_TEMPLATE } from '../../components/github/pull-request-template';
import { Readme } from '../../components/readme';
import { mergeOptions } from '../../utils/merge-options';

function getOptions(options: CdkTypeScriptAppOptions) {
  const { name } = options;

  const defaults = {
    name,
    defaultReleaseBranch: 'main',
    githubOptions: {
      workflows: false,
    },
    gitignore: ['.npmrc', '.vscode'],
    pullRequestTemplateContents: DEFAULT_PULL_REQUEST_TEMPLATE,
    projenrcTs: true,
  } satisfies Partial<CdkTypeScriptAppOptions>;

  return mergeOptions(defaults, options);
}

/**
 * A CDK application in TypeScript
 *
 *
 * @pjid cdk-typescript-app
 */
export class CdkTypeScriptApp extends AwsCdkTypeScriptApp {
  readme: Readme;

  constructor(options: CdkTypeScriptAppOptions) {
    const mergedOptions = getOptions(options);

    super({
      ...mergedOptions,
    } as AwsCdkTypeScriptAppOptions);

    new CodeOwners(this, mergedOptions.codeOwners);
    this.readme = new Readme(this, {
      description: mergedOptions.readme?.description,
    });
    this.readme.addSection(
      'Getting Started',
      '```sh\nyarn install\nnpx projen build\n```',
    );
    if (this.autoMerge) {
      this.github?.mergify?.addRule({
        name: 'Automatic approval for projen upgrade pull requests',
        conditions: [
          'author=endor-projen[bot]',
          ...(this.buildWorkflow?.buildJobIds?.map(
            (id) => `status-success=${id}`,
          ) ?? []),
        ],
        actions: {
          review: {
            type: 'APPROVE',
            message: 'Automatically approving projen upgrade',
          },
        },
      });
      this.github?.mergify?.addRule({
        name: 'Assign PR when check fails',
        conditions: ['#check-failure > 0'],
        actions: {
          assign: {
            add_users: ['daveshepherd'],
          },
        },
      });
    }
  }
}

export * from './cdk-typescript-app-options';
