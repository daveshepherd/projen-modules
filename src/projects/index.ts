export * from './jsii';
export * from './npm';
export * from './python';