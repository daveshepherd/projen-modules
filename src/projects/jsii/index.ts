import { cdk } from 'projen';
import { JsiiProjectOptions } from './jsii-project-options';
import { CodeOwners } from '../../components/github/codeowners';
import { DEFAULT_PULL_REQUEST_TEMPLATE } from '../../components/github/pull-request-template';
import { Readme } from '../../components/readme';
import { mergeOptions } from '../../utils/merge-options';

function getOptions(options: JsiiProjectOptions) {
  const { name } = options;

  const defaults = {
    name,
    autoMerge: true,
    defaultReleaseBranch: 'main',
    gitignore: ['.npmrc', '.vscode'],
    pullRequestTemplateContents: DEFAULT_PULL_REQUEST_TEMPLATE,
    projenrcTs: true,
  } satisfies Partial<JsiiProjectOptions>;

  return mergeOptions(defaults, options);
}

/**
 * A JSII project in TypeScript
 *
 *
 * @pjid jsii-project
 */
export class JsiiProject extends cdk.JsiiProject {
  readme: Readme;

  constructor(options: JsiiProjectOptions) {
    const mergedOptions = getOptions(options);

    super({
      ...mergedOptions,
    } as cdk.JsiiProjectOptions);

    new CodeOwners(this, mergedOptions.codeOwners);
    this.readme = new Readme(this, {
      description: mergedOptions.readme?.description,
    });
    this.readme.addSection(
      'Getting Started',
      `\`\`\`sh\nyarn install\nnpx projen build\n\`\`\`\n
This will:
* Install the dependencies
* Apply any projen changes
* Run tests
* Package project locally

Any files changed by projen should be committed to git.

Running the tests like this will update any snapshot files, this should be reviewed and committed to git.`,
    );
    this.readme.addSection(
      'Testing',
      `Types of testing:
* Snapshot - projen project outputs are stored as a snapshot in the corresponding \`__snapshots__\` directory. When the project changes then it is expected that these snapshots change too and should be reviewed committed alongside the project.
* Unit tests - these assert on specific functionality of the project and should be written for any new functionality added.
`,
    );
    if (this.autoMerge) {
      this.github?.mergify?.addRule({
        name: 'Automatic approval for projen upgrade pull requests',
        conditions: [
          'author=endor-projen[bot]',
          ...(this.buildWorkflow?.buildJobIds?.map(
            (id) => `status-success=${id}`,
          ) ?? []),
        ],
        actions: {
          review: {
            type: 'APPROVE',
            message: 'Automatically approving projen upgrade',
          },
        },
      });
      this.github?.mergify?.addRule({
        name: 'Assign PR when check fails',
        conditions: ['#check-failure > 0'],
        actions: {
          assign: {
            add_users: ['daveshepherd'],
          },
        },
      });
    }
  }
}

export * from './jsii-project-options';
