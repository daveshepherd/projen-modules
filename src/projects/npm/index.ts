import { typescript } from 'projen';
import { NpmPackageOptions } from './npm-package-options';
import { NpmCircleCi } from '../../circleci';
import { CodeOwners } from '../../components/github/codeowners';
import { DEFAULT_PULL_REQUEST_TEMPLATE } from '../../components/github/pull-request-template';
import { Readme } from '../../components/readme';
import { mergeOptions } from '../../utils/merge-options';

function getOptions(options: NpmPackageOptions) {
  const { name } = options;

  const defaults = {
    name,
    defaultReleaseBranch: 'main',
    githubOptions: {
      workflows: false,
    },
    gitignore: ['.npmrc', '.vscode'],
    pullRequestTemplateContents: DEFAULT_PULL_REQUEST_TEMPLATE,
    projenrcTs: true,
  } satisfies Partial<NpmPackageOptions>;

  return mergeOptions(defaults, options);
}

/**
 * A NPM/node package in TypeScript
 *
 *
 * @pjid npm-package
 */
export class NpmPackage extends typescript.TypeScriptProject {
  readme: Readme;

  constructor(options: NpmPackageOptions) {
    const mergedOptions = getOptions(options);

    super({
      ...mergedOptions,
    } as typescript.TypeScriptProjectOptions);

    new CodeOwners(this, mergedOptions.codeOwners);
    new NpmCircleCi(this);
    this.readme = new Readme(this, {
      description: mergedOptions.readme?.description,
    });
    this.readme.addSection(
      'Getting Started',
      '```sh\nyarn install\nnpx projen build\n```',
    );
    if (this.autoMerge) {
      this.github?.mergify?.addRule({
        name: 'Automatic approval for projen upgrade pull requests',
        conditions: [
          'author=endor-projen[bot]',
          ...(this.buildWorkflow?.buildJobIds?.map(
            (id) => `status-success=${id}`,
          ) ?? []),
        ],
        actions: {
          review: {
            type: 'APPROVE',
            message: 'Automatically approving projen upgrade',
          },
        },
      });
    }
  }
}

export * from './npm-package-options';
