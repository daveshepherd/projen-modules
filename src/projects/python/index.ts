import { PythonProject, PythonProjectOptions } from 'projen/lib/python';
import { PythonPackageOptions } from './python-package-options';
import { CodeOwners } from '../../components/github/codeowners';
import { DEFAULT_PULL_REQUEST_TEMPLATE } from '../../components/github/pull-request-template';
import { Readme } from '../../components/readme';
import { mergeOptions } from '../../utils/merge-options';

function getOptions(options: PythonPackageOptions) {
  const { name } = options;

  const defaults = {
    name,
    pullRequestTemplate: true,
    pullRequestTemplateContents: DEFAULT_PULL_REQUEST_TEMPLATE,
  } satisfies Partial<PythonPackageOptions>;

  return mergeOptions(defaults, options);
}

/**
 * A Python package
 *
 *
 * @pjid python-package
 */
export class PythonPackage extends PythonProject {
  readme: Readme;

  constructor(options: PythonPackageOptions) {
    const mergedOptions = getOptions(options);

    super({
      ...mergedOptions,
    } as PythonProjectOptions);

    new CodeOwners(this, mergedOptions.codeOwners);
    if (mergedOptions.pullRequestTemplate ?? true) {
      this.github?.addPullRequestTemplate(
        ...(mergedOptions.pullRequestTemplateContents ?? []),
      );
    }
    this.readme = new Readme(this, {
      description: mergedOptions.readme?.description,
    });
    this.readme.addSection(
      'Getting Started',
      '```sh\nyarn install\nnpx projen build\n```',
    );
  }
}

export * from './python-package-options';
