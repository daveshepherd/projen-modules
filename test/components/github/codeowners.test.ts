import { synthSnapshot } from 'projen/lib/util/synth';
import { JsiiProject } from '../../../src/projects/jsii';
import { NpmPackage } from '../../../src/projects/npm';

describe('CodeOwners', () => {
  it('matches the snapshot with a CODEOWERS file', () => {
    const project = new JsiiProject({
      author: 'Test Person',
      authorAddress: 'test@example.com',
      codeOwners: ['test-team'],
      defaultReleaseBranch: 'main',
      name: 'test',
      repositoryUrl: 'https://github.com/example/example.git',
    });

    const output = synthSnapshot(project);

    expect(output.CODEOWNERS).toContain('* test-team');
  });

  it('matches the snapshot and includes CODEOWNERS in the npmignore file', () => {
    const project = new NpmPackage({
      codeOwners: ['test-team'],
      defaultReleaseBranch: 'main',
      name: 'test',
    });

    const output = synthSnapshot(project);

    expect(output.CODEOWNERS).toMatchSnapshot();
  });
});
