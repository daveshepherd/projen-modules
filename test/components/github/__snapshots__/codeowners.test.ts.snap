// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeOwners matches the snapshot and includes CODEOWNERS in the npmignore file 1`] = `"* @unibuddy/test-team"`;
